sys/kern/event.c

```diff
@@ -1,15 +1,454 @@
+      return ENOENT;
+
+    kn = pool_alloc(P_KNOTE, M_ZERO);
+    kn->kn_kq = kq;
+    kn->kn_kevent = *kev;
+    kn->kn_obj = obj;
+    kn->kn_filtops = filtops;
+
+    SLIST_INSERT_HEAD(knote_list, kn, kn_hashlink);
+
+    if ((error = kn->kn_filtops->filt_attach(kn)) != 0) {
+      knote_drop_detached(kn);
+      return error;
+    }
+  }
+
+  if (kev->flags & EV_DELETE) {
+    knote_drop(kn);
+    return 0;
+  }
+
+  /* `kn_objlock` must be taken to access kn_kevent.udata and
+   * to call filt_event.
+   */
+
+  WITH_MTX_LOCK (kn->kn_objlock) {
+    /*
+     * The user may change some filter values after the
+     * initial EV_ADD, but doing so will not reset any
+     * filter which have already been triggered.
+     */
+    kn->kn_kevent.udata = kev->udata;
+
+    event = kn->kn_filtops->filt_event(kn, 0);
+    WITH_MTX_LOCK (&kq->kq_lock) {
+      if (event && (kn->kn_status & KN_QUEUED) == 0)
+        knote_enqueue(kn);
+    }
+  }
+
+  return 0;
+}
+
+/* Scans for events triggered in a given kqueue. Returns at most `nevents`
+ * events via `eventlist`. If there are no events available, blocks for at
+ * most `tsp`. If `tsp` is null, blocks infinitely. If the timeout is not
+ * positive, returns immediately.
+ */
+static int kqueue_scan(kqueue_t *kq, kevent_t *eventlist, size_t nevents,
+                       timespec_t *tsp, int *retval) {
+  int error, event, timeout;
+  size_t count = 0;
+  systime_t sleepts;
+  knote_tailq_t knqueue;
+  knote_t *kn;
+
+  TAILQ_INIT(&knqueue);
+
+  if (tsp) {
+    timeout = ts2hz(tsp);
+    sleepts = getsystime() + timeout;
+    if (timeout <= 0)
+      timeout = -1; /* don't block */
+  } else {
+    timeout = 0; /* no timeout, wait forever */
+  }
+
+  mtx_lock(&kq->kq_lock);
+
+  /* Block until there are no events or we time out. */
+  while (kq->kq_count == 0) {
+    if (timeout < 0) {
+      error = 0;
+      goto done;
+    }
+
+    error = cv_wait_timed(&kq->kq_cv, &kq->kq_lock, timeout);
+    if (error == EINTR) {
+      mtx_unlock(&kq->kq_lock);
+      return EINTR;
+    }
+
+    if (tsp)
+      timeout = sleepts - getsystime();
+  }
+
+  /* To ensure the correctness of the iteration over pending events,
+   * we need to move already processed knotes to another list. `kq_head`
+   * can change in between iterations, because we are releasing the lock
+   * for a moment.
+   */
+
+  while (count < nevents) {
+    assert(mtx_owned(&kq->kq_lock));
+
+    if (TAILQ_EMPTY(&kq->kq_head))
+      break;
+
+    kn = TAILQ_FIRST(&kq->kq_head);
+    TAILQ_REMOVE(&kq->kq_head, kn, kn_penlink);
+
+    /* To call `filt_event` we need to lock `kn_objlock`, but firstly
+     * we need to release `kq_lock` since holding it would violate the locking
+     * order. `kn_objlock` must be taken before `kq_lock`.
+     */
+    mtx_unlock(&kq->kq_lock);
+    mtx_lock(kn->kn_objlock);
+    event = kn->kn_filtops->filt_event(kn, 0);
+    mtx_lock(&kq->kq_lock);
+    mtx_unlock(kn->kn_objlock);
```

Comment:
This is weird. Is there any particular reason why the lock/unlock don't follow a reverse order? I don't think there's anything wrong with this code, but seeing it written like that makes me think that maybe I'm missing something.

Reply:
I believe that there was not a particular reason and I agree that it may looks confusing.

Changed.